src/Microsoft.Azure.ServiceBus/MessageReceivePump.cs

```diff
@@ -0,0 +1,266 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+namespace Microsoft.Azure.ServiceBus
+{
+    using System;
+    using System.Threading;
+    using System.Threading.Tasks;
+    using Microsoft.Azure.ServiceBus.Primitives;
+
+    public sealed class MessageReceivePump
+    {
+        const int MaxInitialReceiveRetryCount = 3;
+        static readonly TimeSpan ServerBusyExceptionBackoffAmount = TimeSpan.FromSeconds(10);
+        static readonly TimeSpan OtherExceptionBackoffAmount = TimeSpan.FromSeconds(1);
+        readonly Func<BrokeredMessage, CancellationToken, Task> onMessageCallback;
+        readonly OnMessageOptions onMessageOptions;
+        readonly MessageReceiver messageReceiver;
+        readonly CancellationToken pumpCancellationToken;
+        readonly SemaphoreSlim maxConcurrentCallsSemaphoreSlim;
+
+        public MessageReceivePump(
+            MessageReceiver messageReceiver,
+            OnMessageOptions onMessageOptions,
+            Func<BrokeredMessage, CancellationToken, Task> callback,
+            CancellationToken pumpCancellationToken)
+        {
+            if (messageReceiver == null)
+            {
+                throw new ArgumentNullException(nameof(messageReceiver));
+            }
+
+            this.messageReceiver = messageReceiver;
+            this.onMessageOptions = onMessageOptions;
+            this.onMessageCallback = callback;
+            this.pumpCancellationToken = pumpCancellationToken;
+            this.maxConcurrentCallsSemaphoreSlim = new SemaphoreSlim(this.onMessageOptions.MaxConcurrentCalls);
+        }
+
+        public async Task StartPumpAsync()
+        {
+            int retryCount = 0;
+            BrokeredMessage initialMessage = null;
+            while (true)
+            {
+                try
+                {
+                    initialMessage = await this.messageReceiver.ReceiveAsync().ConfigureAwait(false);
+                    MessagingEventSource.Log.MessageReceiverPumpInitialMessageReceived(this.messageReceiver.ClientId, initialMessage);
+                    break;
+                }
+                catch (Exception exception)
+                {
+                    retryCount++;
+                    MessagingEventSource.Log.MessageReceiverPumpInitialMessageReceiveException(this.messageReceiver.ClientId, retryCount, exception);
+
+                    if (retryCount == MaxInitialReceiveRetryCount ||
+                        !this.ShouldRetry(exception))
+                    {
+                        throw;
+                    }
+
+                    TimeSpan backOffTime = this.GetBackOffTime(exception);
+                    await Task.Delay(backOffTime, this.pumpCancellationToken).ConfigureAwait(false);
+                }
+            }
+
+            TaskExtensionHelper.Schedule(() => this.MessagePumpTask(initialMessage));
```

Comment:
Even with the way you suggest, it is still making an additional call right? I don't see how this saves an additional call.. Or are you talking about readability here? Besides, with the way Schedule is written wouldn't the JIT compiler inline it anyway, so it wouldn't result in an extra call everytime